packages/contract-call-decoder/src/lib/ContractCallDecoder.ts

```diff
@@ -0,0 +1,191 @@
+import radspec from '@blossom-labs/rosette-radspec';
+import { Transaction as TransactionRosette } from '@blossom-labs/rosette-radspec/dist/declarations/src/types/web3';
+import { decode as decodeBytecode } from '@ethereum-sourcify/bytecode-utils';
+import { FunctionFragment, Interface } from '@ethersproject/abi';
+import { EthereumProvider } from 'ethereum-provider';
+
+import { extractCustomFields, getValueFromDecodedFunctionData } from './utils';
+
+type Transaction = TransactionRosette & {
+  readonly chainId?: number;
+};
+
+require('isomorphic-fetch');
+
+export enum MetadataSources {
+  Sourcify,
+  BytecodeMetadata,
+}
+
+type GetMetadataOptions = {
+  readonly source?: MetadataSources;
+  readonly chainId?: number;
+  readonly address?: string;
+  readonly rpcProvider?: EthereumProvider;
+  readonly ipfsGateway?: string;
+  readonly sourcifyProvider?: string;
+};
+
+const defaultGetMetadataOptions: GetMetadataOptions = {
+  source: MetadataSources.Sourcify,
+  sourcifyProvider: 'https://repo.sourcify.dev',
+  ipfsGateway: 'https://cloudflare-ipfs.com/',
+};
+
+export async function getMetadataFromAddress(options: GetMetadataOptions) {
+  options = { ...defaultGetMetadataOptions, ...options };
+  // eslint-disable-next-line functional/no-let
+  let contractMetadataJSON;
+  if (options.source === MetadataSources.Sourcify) {
+    try {
+      const req = await fetch(
+        `${options.sourcifyProvider}/contracts/full_match/${options.chainId}/${options.address}/metadata.json`
+      );
+      contractMetadataJSON = await req.json();
+    } catch (e) {
+      console.log(e);
+      return false;
+    }
+  } else if (options.source === MetadataSources.BytecodeMetadata) {
+    const bytecode = (await options.rpcProvider.request({
+      method: 'eth_getCode',
+      params: [options.address, 'latest'],
+    })) as string;
+    if (!bytecode || bytecode === '0x') {
+      return false;
+    }
+    const { ipfs: metadataIpfsCid } = decodeBytecode(bytecode);
+    try {
+      const req = await fetch(`${options.ipfsGateway}/ipfs/${metadataIpfsCid}`);
+      contractMetadataJSON = await req.json();
+    } catch (e) {
+      console.log(e);
+      return false;
+    }
+  }
+
+  return contractMetadataJSON;
+}
+
+export const evaluate = async function name(
+  expression: string,
+  abi,
+  transaction: Transaction,
+  provider?
+): Promise<string> {
+  return await radspec(expression, abi, transaction, provider);
```

Comment:
I see `provider` in `radspec` is mandatory but optional in the function param. Also `provider` and `abi` are `any` type here. Shouldn't this be not allowed in TS?